rust/xaynet-core/src/mask/masking.rs

```diff
@@ -897,4 +1013,135 @@ mod tests {
     test_masking_and_aggregation!(pow_i64_b4, Power2, i64, 10_000, 10, 5);
     test_masking_and_aggregation!(pow_i64_b6, Power2, i64, 1_000_000, 10, 5);
     test_masking_and_aggregation!(pow_i64_bmax, Power2, i64, 10, 5);
+
+    /// Generate tests for masking, aggregation and unmasking of multiple models:
+    /// - generate random scalars from a uniform distribution with a seeded PRNG
+    /// - scale a model of unit weights, mask and aggregate it to the aggregated masked models
+    /// - derive a mask from the mask seed and aggregate it to the aggregated masks
+    /// - unmask the aggregated masked model
+    /// - check that all aggregated unmasked weights are equal to the original unit weights (up
+    ///   to a tolerance determined by the masking configuration)
+    ///
+    /// The arguments to the macro are:
+    /// - a suffix for the test name
+    /// - the group type of the model and scalar (variants of `GroupType`)
+    /// - the data type of the model and scalar (either float primitives or float variants of
+    ///   `DataType`)
+    /// - an absolute bound for the scalar (optional, choices: 1, 100, 10_000, 1_000_000)
+    /// - the number of weights per model
+    /// - the number of models
+    macro_rules! test_masking_and_aggregation_scalar {
+        ($suffix:ident, $group:ty, $data:ty, $bound:expr, $len:expr, $count:expr $(,)?) => {
+            paste::item! {
+                #[test]
+                fn [<test_masking_and_aggregation_scalar $suffix>]() {
+                    // Step 1: Build the masking config
+                    let config = MaskConfig {
+                        group_type: $group,
+                        data_type: paste::expr! { [<$data:upper>] },
+                        bound_type: match $bound {
+                            1 => B0,
+                            100 => B2,
+                            10_000 => B4,
+                            1_000_000 => B6,
+                            _ => Bmax,
+                        },
+                        model_type: M3,
+                    };
+                    let vect_len = $len as usize;
+
+                    // Step 2: Generate random scalars
+                    // take vectors [1, ..., 1] as models to scale
+                    let bound = if $bound == 0 {
```

Comment:
It seems that `$bound` is never 0, so is this really needed?

Reply:
it is used for the `Bmax` cases in the 2nd branch of the macro.